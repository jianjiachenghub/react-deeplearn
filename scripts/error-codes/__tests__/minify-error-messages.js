/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
/* eslint-disable quotes */
'use strict';

let babel = require('babel-core');
let devExpressionWithCodes = require('../minify-error-messages');

function transform(input) {
  return babel.transform(input, {
    plugins: [devExpressionWithCodes],
  }).code;
}

let oldEnv;

describe('error transform', () => {
  beforeEach(() => {
    oldEnv = process.env.NODE_ENV;
    process.env.NODE_ENV = '';
  });

  afterEach(() => {
    process.env.NODE_ENV = oldEnv;
  });

  it('should replace simple invariant calls', () => {
    expect(
      transform(`
import invariant from 'shared/invariant';
invariant(condition, 'Do not override existing functions.');
`)
    ).toMatchSnapshot();
  });

  it('should only add `ReactError` and `ReactErrorProd` once each', () => {
    expect(
      transform(`
import invariant from 'shared/invariant';
invariant(condition, 'Do not override existing functions.');
invariant(condition, 'Do not override existing functions.');
`)
    ).toMatchSnapshot();
  });

  it('should support invariant calls with args', () => {
    expect(
      transform(`
import invariant from 'shared/invariant';
invariant(condition, 'Expected %s target to be an array; got %s', foo, bar);
`)
    ).toMatchSnapshot();
  });

  it('should support invariant calls with a concatenated template string and args', () => {
    expect(
      transform(`
import invariant from 'shared/invariant';
invariant(condition, 'Expected a component class, ' + 'got %s.' + '%s', Foo, Bar);
`)
    ).toMatchSnapshot();
  });

  it('should correctly transform invariants that are not in the error codes map', () => {
    expect(
      transform(`
import invariant from 'shared/invariant';
invariant(condition, 'This is not a real error message.');
`)
    ).toMatchSnapshot();
  });

  it('should handle escaped characters', () => {
    expect(
      transform(`
import invariant from 'shared/invariant';
invariant(condition, 'What\\'s up?');
`)
    ).toMatchSnapshot();
  });
});
