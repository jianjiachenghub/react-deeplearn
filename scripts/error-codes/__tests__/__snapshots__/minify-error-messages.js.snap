// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`error transform should correctly transform invariants that are not in the error codes map 1`] = `
"import _ReactError from 'shared/ReactError';

import invariant from 'shared/invariant';
(function () {
  if (!condition) {
    throw _ReactError(\`This is not a real error message.\`);
  }
})();"
`;

exports[`error transform should handle escaped characters 1`] = `
"import _ReactError from 'shared/ReactError';

import invariant from 'shared/invariant';
(function () {
  if (!condition) {
    throw _ReactError(\`What's up?\`);
  }
})();"
`;

exports[`error transform should only add \`ReactError\` and \`ReactErrorProd\` once each 1`] = `
"import _ReactErrorProd from 'shared/ReactErrorProd';
import _ReactError from 'shared/ReactError';

import invariant from 'shared/invariant';
(function () {
  if (!condition) {
    if (__DEV__) {
      throw _ReactError(\`Do not override existing functions.\`);
    } else {
      throw _ReactErrorProd(16);
    }
  }
})();
(function () {
  if (!condition) {
    if (__DEV__) {
      throw _ReactError(\`Do not override existing functions.\`);
    } else {
      throw _ReactErrorProd(16);
    }
  }
})();"
`;

exports[`error transform should replace simple invariant calls 1`] = `
"import _ReactErrorProd from 'shared/ReactErrorProd';
import _ReactError from 'shared/ReactError';

import invariant from 'shared/invariant';
(function () {
  if (!condition) {
    if (__DEV__) {
      throw _ReactError(\`Do not override existing functions.\`);
    } else {
      throw _ReactErrorProd(16);
    }
  }
})();"
`;

exports[`error transform should support invariant calls with a concatenated template string and args 1`] = `
"import _ReactErrorProd from 'shared/ReactErrorProd';
import _ReactError from 'shared/ReactError';

import invariant from 'shared/invariant';
(function () {
  if (!condition) {
    if (__DEV__) {
      throw _ReactError(\`Expected a component class, got \${Foo}.\${Bar}\`);
    } else {
      throw _ReactErrorProd(18, Foo, Bar);
    }
  }
})();"
`;

exports[`error transform should support invariant calls with args 1`] = `
"import _ReactErrorProd from 'shared/ReactErrorProd';
import _ReactError from 'shared/ReactError';

import invariant from 'shared/invariant';
(function () {
  if (!condition) {
    if (__DEV__) {
      throw _ReactError(\`Expected \${foo} target to be an array; got \${bar}\`);
    } else {
      throw _ReactErrorProd(7, foo, bar);
    }
  }
})();"
`;
